src/message.rs

```diff
@@ -145,6 +148,25 @@ macro_rules! message_args_impl {
                 })
             }
 
+            #[cfg(all(feature="gnustep_runtime",
+                      not(any(target_arch = "arm",
+                              target_arch = "x86",
+                              target_arch = "x86_64"))))]
+            unsafe fn send<T, R>(self, obj: *mut T, sel: Sel) -> R
+                   where T: Message, R: Any {
+                    let mut receiver = obj as *mut Object;
+                    let nil: *mut Object = ::std::ptr::null_mut();
+                    let ref slot = *runtime::objc_msg_lookup_sender(&mut receiver as *mut *mut Object, sel, nil);
```

Comment:
We can only use `objc_msgSend` on the platforms quoted in the earlier comment (hence it's conditional on the target arch). For everything else, we need to fall back to this mechanism which looks up the slot first and then calls the IMP.  

You're right that the slot mechanism is immensely useful for caching IMPs, but I don't see how to build a platform-independent abstraction over this. The reason is that caching IMPs is generally unsafe on the Apple runtime. Things like isa-swizzling, `method_exchangeImplementation` or simply adding a KVO observer might leave you with an outdated IMP and no way to detect it.
With the GNUstep runtime, the slot version will be incremented, giving you a way to invalidate stale cached IMPs.